### PR TITLE
Add HTTP api for TTN

### PR DESCRIPTION
**Pull request by romeovs**
This commit contains work to create a HTTP api for TTN. Refer to the `api` folder for more info.

### CODE REVIEW

File: api/auth/main.go [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			return nil, errors.New("invalid token algorithm")
+		} else {
+			return key, nil
+		}
+	}
+}
+
+// validate an access token
+func ValidateToken (server, accessToken string) (map[string]interface{}, error) {
+
```
Comment (johanstokking) [inline api/auth/main.go]: Remove unnecessary spacing

File: api/models/application.go [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+type app_ Application
+func (app Application) MarshalJSON() ([]byte, error) {
+	return json.Marshal(struct {
+			app_
+			Type  string `json:"@type"`
+			Self  string `json:"@self"`
+		}{
+			app_:  app_(app),
+			Type: "application",
+			Self: fmt.Sprintf("/api/v1/applications/%X", app.EUI),
```
Comment (johanstokking) [inline api/models/application.go]: Leave for now

File: api/models/application.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+	for _, app := range apps {
+		if bytes.Equal(app.EUI, eui) {
+			return app, nil
+		}
+	}
+	return app, errors.New("Application with that eui does not exist")
+}
+
+
+func CreateApplication(accessToken auth.Token, eui EUI, name string) (app Application, err error) {
```
Comment (johanstokking) [inline api/models/application.go]: Introduce `AppEUI` and `DevEUI` type alias in `core` package

File: api/models/application.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+	return app, nil
+}
+
+
+// Authorizes user to application
+// TODO: should I check wether the user exists?
+func AuthorizeUserToApplication(accessToken auth.Token, eui EUI, email string) error {
+
+	authServer := viper.GetString("ttn-account-server")
+	uri        := fmt.Sprintf("%s/applications/%s/authorize", authServer, WriteEUI(eui))
```
Comment (johanstokking) [inline api/models/application.go]: Replace `WriteEUI` for `fmt.Sprintf("%X", ...)`

File: api/models/application.go [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+	if err != nil {
+		return err
+	}
+
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+
+	defer resp.Body.Close()
```
Comment (johanstokking) [inline api/models/application.go]: Only close body when you use the `Body`

File: api/models/application.go [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+		return app, err
+	}
+
+	// yay! nothing went wrong
+	return app, nil
+}
+
+
+// Authorizes user to application
+// TODO: should I check wether the user exists?
```
Comment (johanstokking) [inline api/models/application.go]: Nop

File: api/models/application.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+	req, err := auth.NewAuthRequest(accessToken, "DELETE", uri, nil)
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+
+	defer resp.Body.Close()
```
Comment (johanstokking) [inline api/models/application.go]: See above

File: api/models/devices.go [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+	"github.com/TheThingsNetwork/ttn/core"
+	"github.com/TheThingsNetwork/ttn/core/components/handler"
+	"github.com/TheThingsNetwork/ttn/utils/random"
+	"github.com/TheThingsNetwork/ttn/api/auth"
+	"golang.org/x/net/context"
+	"encoding/json"
+)
+
+type Device struct {
+	DevEUI   EUI     `json:"eui,omitempty"`
```
Comment (johanstokking) [inline api/models/devices.go]: `devEUI`

File: api/models/devices.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+	AppEUI   EUI     `json:"app"`
```
Comment (johanstokking) [inline api/models/devices.go]: `appEUI`

File: api/models/devices.go [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+	Type     string  `json:"type"`
+	Address  []byte  `json:"address"`
+	FCntUp   uint32  `json:"fcnt_up"`
+	FCntDown uint32  `json:"fcnt_down"`
+	Active   bool    `json:"activated"`
+	AppSKey  SKey    `json:"app_secret"`
+	NwkSKey  SKey    `json:"nwk_secret"`
```
Comment (johanstokking) [inline api/models/devices.go]: Introduce type alias for both types of keys

Use JSON field names that match field names of structure

File: api/models/devices.go [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+	"github.com/TheThingsNetwork/ttn/api/auth"
+	"golang.org/x/net/context"
+	"encoding/json"
+)
+
+type Device struct {
+	DevEUI   EUI     `json:"eui,omitempty"`
+	AppEUI   EUI     `json:"app"`
+	Type     string  `json:"type"`
+	Address  []byte  `json:"address"`
```
Comment (johanstokking) [inline api/models/devices.go]: `DevAddr`

File: api/models/devices.go [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+		return nil, err
+	}
+
+	result := make([]Device, 0, len(res.ABP) + len(res.OTAA))
+
+	for i, device := range res.ABP {
+		result[i] = Device{
+			DevEUI:   nil,
+			AppEUI:   appEui,
+			Type:     "personal",
```
Comment (johanstokking) [inline api/models/devices.go]: `personalized`

File: api/models/devices.go [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+			NwkSKey:  device.NwkSKey,
+		}
+	}
+
+	n := len(res.ABP)
+	for i, device := range res.OTAA {
+		result[i + n] = Device{
+			DevEUI:   device.DevEUI,
+			AppEUI:   appEui,
+			Type:     "dynamic",
```
Comment (johanstokking) [inline api/models/devices.go]: `otaa`

File: api/models/devices.go [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+		case 0:
+			// allow for random appKey
+			appKey = SKey(random.Bytes(16))
+		case 1:
+			// get appKey from param
+			appKey = rest[0]
+
+		default:
+			return errors.New("illegal amount of parameters passed to RegisterOTAA")
+	}
```
Comment (johanstokking) [inline api/models/devices.go]: Pass `nil` for no AppKey

File: api/models/subscribe.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+func ConnectMQTTClient(token auth.Token, appEUI EUI) (client mqtt.Client, err error) {
+	broker := fmt.Sprintf("tcp://%s", viper.GetString("mqtt-broker"))
+
+	app, err := GetApplication(token, appEUI)
+	if err != nil {
+		return client, err
+	}
+
+	client = mqtt.NewClient(nil, "ttnhttp", WriteEUI(app.EUI), string(app.AccessKeys[0]), broker)
```
Comment (johanstokking) [inline api/models/subscribe.go]: `dashboard`, check what it does. Probably prefix for client ID

File: api/router/devices.go [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+type registerDeviceFields struct {
+	AppEUI  models.EUI  `json:"application"`
+	DevEUI  models.EUI  `json:"device"`
+	AppKey  models.SKey `json:"app_secret" binding:"required"`
+	NwkKey  models.SKey `json:"nwk_secret"`
+	Type    string      `json:"type" binding:"required"`
+	Address []byte      `json:"address"`
+}
+
+func registerDevice(c *gin.Context) {
```
Comment (johanstokking) [inline api/router/devices.go]: Use different endpoints for OTAA and personalized, also with different JSON contracts. Fields are very different

File: api/models/devices.go (hunk 12 above)
Comment (htdvisser) [inline api/models/devices.go]: not `abp`?

File: api/models/subscribe.go (hunk 15 above)
Comment (htdvisser) [inline api/models/subscribe.go]: Indeed prefix

File: api/router/applications.go [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+// get application by eui
+func getApplication (c *gin.Context) {
+	token, authorized := c.Get("token");
+	if !authorized {
+		c.AbortWithStatus(401)
+		return
+	}
+
+	// EUI as string
```
Comment (htdvisser) [inline api/router/applications.go]: This comment is not necessary

File: api/router/applications.go [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+}
+
+func getDevicesForApp (c *gin.Context) {
+	token, authorized := c.Get("token");
+	if !authorized {
+		c.AbortWithStatus(401)
+		return
+	}
+
+	// EUI as string
```
Comment (htdvisser) [inline api/router/applications.go]: This comment is not necessary

File: api/router/applications.go [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+	eui, _ := c.Get("appEUI")
+
+	// get the app
```
Comment (htdvisser) [inline api/router/applications.go]: This comment is not necessary

File: api/models/devices.go [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+// Fetch devices for specific app
+func DevicesForApp (accessToken auth.Token, appEui EUI) ([]Device, error) {
+	println(string(accessToken))
+	manager := getHandlerManager()
+	res, err := manager.ListDevices(context.Background(), &core.ListDevicesHandlerReq{
+		Token:  string(accessToken),
+		AppEUI: []byte(appEui),
+	})
+
+	if err != nil {
```
Comment (htdvisser) [inline api/models/devices.go]: Here, a "Not Found" should be handled as `[]`

File: api/router/applications.go [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+		c.AbortWithError(404, err)
+		return
+	}
+
+	devices, err := models.DevicesForApp(token.(auth.Token), app.EUI)
+	if err != nil {
+		devices = make([]models.Device, 0)
+		// c.JSON(200, de)
+		// c.AbortWithError(500, err)
+		// return
```
Comment (htdvisser) [inline api/router/applications.go]: `// I couldn't find the delete key`?

File: api/router/applications.go [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+
+	c.JSON(200, devices)
+}
+
+type createAppFields struct {
+	Name string     `json:"name" binding:"required"`
+	Eui  models.EUI `json:"eui"`
+}
+
+// creates a new application
```
Comment (htdvisser) [inline api/router/applications.go]: I think `createApplication` doesn't need a comment explaining what it does

File: api/router/subscribe.go [hunk 23]
```diff
@@ -0,0 +1,54 @@
+package router
```
Comment (htdvisser) [inline api/router/subscribe.go]: Maybe just delete this file

**Pull request closed without merging**